toolchain/check/testdata/facet/validate_rewrite_constraints.carbon

```diff
@@ -0,0 +1,1353 @@
+  // CHECK:STDERR:   F({}, C);
+  // CHECK:STDERR:   ^~~~~~~~
+  // CHECK:STDERR: fail_generic_interface_facet_value_wrong_specific_impl.carbon:[[@LINE-7]]:1: note: while deducing parameters of generic declared here [DeductionGenericHere]
+  // CHECK:STDERR: fn F(U:! type, T:! I(U) where .X = ()) {}
+  // CHECK:STDERR: ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  F({}, C);
+}
+
+// --- dependent_rules.carbon
+library "[[@TEST_NAME]]";
+
+interface I { let X:! type; }
+interface J { let Y:! type; }
+
+fn F(T:! I & J where .X = .Y) {
+  // Allowed since `T impls I`, `T impls J`, and has constraint providing
+  // T.(I.X) = T.(J.Y)`.
+  F(T);
+}
+
+// --- fail_convert.carbon
+library "[[@TEST_NAME]]";
+
+interface I { let X:! type; }
+
+fn F(T:! I where .X = {.a: (), .b: {}}) {}
+
+fn H() {
+  class C;
+  impl C as I where .X = {.b: {}, .a: ()} {}
+  // CHECK:STDERR: fail_convert.carbon:[[@LINE+7]]:3: error: cannot convert type `C` into type implementing `I where .(I.X) = {.a: (), .b: {}}` [ConversionFailureTypeToFacet]
+  // CHECK:STDERR:   F(C);
+  // CHECK:STDERR:   ^~~~
+  // CHECK:STDERR: fail_convert.carbon:[[@LINE-8]]:6: note: initializing generic parameter `T` declared here [InitializingGenericParam]
+  // CHECK:STDERR: fn F(T:! I where .X = {.a: (), .b: {}}) {}
+  // CHECK:STDERR:      ^
+  // CHECK:STDERR:
+  F(C);
+}
+
+fn G(T:! I where .X = {.b: {}, .a: ()}) {
+  // CHECK:STDERR: fail_convert.carbon:[[@LINE+7]]:3: error: cannot convert type `T` that implements `I where .(I.X) = {.b: {}, .a: ()}` into type implementing `I where .(I.X) = {.a: (), .b: {}}` [ConversionFailureFacetToFacet]
+  // CHECK:STDERR:   F(T);
+  // CHECK:STDERR:   ^~~~
+  // CHECK:STDERR: fail_convert.carbon:[[@LINE-19]]:6: note: initializing generic parameter `T` declared here [InitializingGenericParam]
+  // CHECK:STDERR: fn F(T:! I where .X = {.a: (), .b: {}}) {}
+  // CHECK:STDERR:      ^
+  // CHECK:STDERR:
+  F(T);
+}
+
+// --- fail_todo_dependent_rules_compound.carbon
+library "[[@TEST_NAME]]";
+
+interface I { let X:! type; }
+interface J { let Y:! type; }
+
+// CHECK:STDERR: fail_todo_dependent_rules_compound.carbon:[[@LINE+8]]:23: error: expected identifier or `Self` after `.` [ExpectedIdentifierOrSelfAfterPeriod]
+// CHECK:STDERR: fn F(T:! I & J where .(I.X) = .(J.Y)) {
+// CHECK:STDERR:                       ^
+// CHECK:STDERR:
+// CHECK:STDERR: fail_todo_dependent_rules_compound.carbon:[[@LINE+4]]:23: error: semantics TODO: `handle invalid parse trees in `check`` [SemanticsTodo]
+// CHECK:STDERR: fn F(T:! I & J where .(I.X) = .(J.Y)) {
+// CHECK:STDERR:                       ^
+// CHECK:STDERR:
+fn F(T:! I & J where .(I.X) = .(J.Y)) {
+  // Allowed since `T impls I`, `T impls J`, and has constraint providing
+  // T.(I.X) = T.(J.Y)`.
+  F(T);
+}
+
+// --- parameterized_interface.carbon
+library "[[@TEST_NAME]]";
+
+interface I(T:! type) { let X:! type; }
+
+final impl forall [J:! I(())] J as I({}) where .X = {} {}
```

Comment:
Should this one be `forall [J:! I(()) where .X = ()]`? You provide a `where .X = ()` below but it seems to not be relevant.

Reply:
Ah, yes, good point. Then we're verifying we pick up the `.X` from the right place and the 2nd one actually does something.